### ISSUE TITLE
Low performance of trained classfier when the input labels are clustered

### ISSUE THREAD

**Issue by xiamx** [facebookresearch/fastText#74]
When training a classifier, if the input contain labels that are clustered, the precision and recall of the trained model is very low e.g.

```
some text __label__a
other text __label__a
yet another line __label__b
the forth line __label__b
the fifth line __label__c
last line __label__c
```

If the input lines are shuffled, the precision and recall of the trained model are higher. 
Is this an expected behavior?

**Comment by EdouardGrave** [comment]
Yes, this is an expected behavior. FastText uses stochastic gradient descent to learn the model, and the examples are processed "in order". Thus, it is important to shuffle the training data, so that not all the examples of a given class are seen first, then the examples of a second class, etc.

**Comment by riyadparvez** [comment]
Wouldn't it be easier to just shuffle training data before training in fastText itself so that users do not get surprised?

**Comment by adam2326** [comment]
No since fastText does not accept IDs separately it would be difficult to join data back if it was part of a larger data set.  I dont want to append an ID to the text itself.  So shuffle-> break text off -> fasttext -> join.  I can count on order this way.
